docs/decisions/0033-rag-in-sk.md

```diff
@@ -0,0 +1,297 @@
+
+This approach doesn't include any memory connectors in Semantic Kernel out-of-the-box, but at the same time it gives an opportunity for developers to handle their data in the way that works for them the best.
+
+String concatenation:
+
+```csharp
+var kernel = Kernel.CreateBuilder()
+    .AddOpenAIChatCompletion("model-id", "api-key")
+    .Build();
+
+var builder = new StringBuilder();
+
+// User is responsible for searching the data in a way of their choice, this is an example how it could look like.
+var data = await this._vectorDB.SearchAsync("Company budget by year");
+
+builder.AppendLine(data);
+builder.AppendLine("What is my budget for 2024?");
+
+var result = await kernel.InvokePromptAsync(builder.ToString());
+```
+
+Prompt template and Kernel arguments:
+
+```csharp
+var kernel = Kernel.CreateBuilder()
+    .AddOpenAIChatCompletion("model-id", "api-key")
+    .Build();
+
+// User is responsible for searching the data in a way of their choice, this is an example how it could look like.
+var data = await this._vectorDB.SearchAsync("Company budget by year");
+
+var arguments = new KernelArguments { ["budgetByYear"] = data };
+
+var result = await kernel.InvokePromptAsync("{{budgetByYear}} What is my budget for 2024?", arguments);
+```
+
+### Option 2 [Supported] - Memory as Plugin
+
+This approach is similar to Option 1, but data search step is part of prompt rendering process. Following list contains possible plugins to use for data search:
+
+- [ChatGPT Retrieval Plugin](https://github.com/openai/chatgpt-retrieval-plugin) - this plugin should be hosted as a separate service. It has integration with various [vector databases](https://github.com/openai/chatgpt-retrieval-plugin?tab=readme-ov-file#choosing-a-vector-database).
+- [SemanticKernel.Plugins.Memory.TextMemoryPlugin](https://www.nuget.org/packages/Microsoft.SemanticKernel.Plugins.Memory) - Semantic Kernel solution, which supports various [vector databases](https://learn.microsoft.com/en-us/semantic-kernel/memories/vector-db#available-connectors-to-vector-databases).
+- Custom user plugin.
+
+ChatGPT Retrieval Plugin:
+
+```csharp
+var kernel = Kernel.CreateBuilder()
+    .AddOpenAIChatCompletion("model-id", "api-key")
+    .Build();
+
+// Import ChatGPT Retrieval Plugin using OpenAPI specification
+// https://github.com/openai/chatgpt-retrieval-plugin/blob/main/.well-known/openapi.yaml
+await kernel.ImportPluginFromOpenApiAsync("ChatGPTRetrievalPlugin", openApi!, executionParameters: new(authCallback: async (request, cancellationToken) =>
+{
+    request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", "chat-gpt-retrieval-plugin-token");
+}));
+
+const string Query = "What is my budget for 2024?";
+const string Prompt = "{{ChatGPTRetrievalPlugin.query_query_post queries=$queries}} {{$query}}";
```

Comment:
The ChatGPT retrieval plugin has different functions (name and behavior) when compared with our `TextMemoryPlugin`. Would it be reasonable to align so that developers could easily swap? The ChatGPT retrieval plugin has the following functions:

- `upsert_file`
- `upsert`
- `query`
- `delete`

https://github.com/openai/chatgpt-retrieval-plugin/blob/main/server/main.py

Reply:
While ChatGPT retrieval plugin contains methods to upsert, query and delete data, the OpenAPI specification, which we are using to import it as plugin, contains only `query` method described:
https://github.com/openai/chatgpt-retrieval-plugin/blob/main/.well-known/openapi.yaml

So, in order to make upsert and delete methods availabile, we will have to modify this OpenAPI specification and add remaining methods to it.

Reply:
Maybe put another way... if it's called `query` in the ChatGPT plugin, is there an opportunity to rename our `recall` function to `query`? That way, a user could "swap" without having to change any of their prompts (as long as the maintained the same plugin name during import)?

Reply:
Also, it looks like they do have an example of how to expose the other endpoints (e.g., upsert for memory), so modifying the OpenAPI specification doesn't seem outside the realm of possibility.

Reply:
> Maybe put another way... if it's called `query` in the ChatGPT plugin, is there an opportunity to rename our `recall` function to `query`? That way, a user could "swap" without having to change any of their prompts (as long as the maintained the same plugin name during import)?

Yes, sorry, I misunderstood your initial question. I think it makes sense to rename `recall` method in current `TextMemoryPlugin` or create new plugin that will use `query` method name or something similar - `retrieve`, `search` etc.